### PR TITLE
Migrate localization page

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrates "Improve > International > Localization" page to Symfony
| Type?         | new feature
| Category?     | CO
| BC breaks?    | yes (overrides)
| Deprecations? | no
| Fixed ticket? | n/a
| How to test?  | Access "Improve > International > Localization" and it should work the same as legacy page.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9116)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
@sarjon this need a rebase and probably some refactoring, I'll review tomorrow :)

File: src/PrestaShopBundle/Resources/views/Admin/Improve/International/Localization/Blocks/configuration.html.twig
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/International/Localization/Blocks/configuration.html.twig]: _The default currency used in your shop._ must be attached to the Admin.International.Help domains.

Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Improve/International/Localization/Blocks/configuration.html.twig]: yup, thanks. :)

Comment (LouiseBonnard) [reply src/PrestaShopBundle/Resources/views/Admin/Improve/International/Localization/Blocks/configuration.html.twig]: You're welcome, Šarūnas!

**Comment by mickaelandrieu** [comment]
Hello @rokaszygmantas, this need a rebase :)

File: src/Core/Localization/Pack/LocalizationPackLoader.php
Comment (mickaelandrieu) [inline src/Core/Localization/Pack/LocalizationPackLoader.php]: wrong header

Comment (sarjon) [reply src/Core/Localization/Pack/LocalizationPackLoader.php]: oh gosh... :man_facepalming:

File: src/Core/Localization/Pack/LocalizationPackDownloader.php
Comment (sarjon) [inline src/Core/Localization/Pack/LocalizationPackDownloader.php]: not sure if it's the best idea to add another method to ```Tools``` adapter class, what do you think @mickaelandrieu ?

File: src/Adapter/Tools.php
Comment (PierreRambaud) [inline src/Adapter/Tools.php]: Use namespace please, `LegacyTools` is available :)

File: src/Core/Localization/Pack/LocalizationPackDownloader.php
Comment (mickaelandrieu) [reply src/Core/Localization/Pack/LocalizationPackDownloader.php]: why do we need something better than the native PHP function at first place? :/

Comment (sarjon) [reply src/Core/Localization/Pack/LocalizationPackDownloader.php]: behind the scene (inside legacy `Tools::file_get_contents`) it actually uses `curl_init` function and some additional configuration to get localization data, so I don't want to duplicate it. 
On the other hand, I tested it with native `file_get_contents` and it works fine. :thinking:

File: src/Adapter/Tools.php
Comment (mickaelandrieu) [inline src/Adapter/Tools.php]: we can't remove it, it's part of public API. But you can deprecate it.

File: src/Core/Localization/Pack/Factory/LocalizationPackFactoryInterface.php
Comment (mickaelandrieu) [inline src/Core/Localization/Pack/Factory/LocalizationPackFactoryInterface.php]: you cant' use object model directly in Core classes. Rely on an Adapter instead.

File: src/Adapter/Tools.php
Comment (sarjon) [reply src/Adapter/Tools.php]: it's not part of public API, I have added it and now removed as it's not needed anymore.

File: src/Core/Localization/Pack/Factory/LocalizationPackFactoryInterface.php
Comment (sarjon) [reply src/Core/Localization/Pack/Factory/LocalizationPackFactoryInterface.php]: yup. :) I found that there's adapter aliases for legacy object models under `PrestaShop\PrestaShop\Adapter\Entity` namespace, is it okay if i use it?

Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Factory/LocalizationPackFactoryInterface.php]: Yes, thanks

**Comment by sarjon** [comment]
@mickaelandrieu this needs to be reviewed. :)

File: src/Adapter/Language/LanguageManager.php
Comment (mickaelandrieu) [inline src/Adapter/Language/LanguageManager.php]: How about LanguageActivator instead? Let's remove every "manager", these classes have multiples responsabilities and we don't want it anymore.

File: src/Adapter/Localization/LocalUnitsConfiguration.php
Comment (mickaelandrieu) [inline src/Adapter/Localization/LocalUnitsConfiguration.php:71]: it's not related to this pull request, but shouldn't we add a generic error when the configuration is invalid?

File: src/Adapter/Language/LanguageManager.php
Comment (mickaelandrieu) [reply src/Adapter/Language/LanguageManager.php]: Also would you mind to create LanguageActivatorInterface in `Core/Language/` so when we will remove the adapter we can rely on the interface :)

File: src/Core/Localization/Pack/Import/LocalizationPackImportConfig.php
Comment (mickaelandrieu) [inline src/Core/Localization/Pack/Import/LocalizationPackImportConfig.php]: this needs an interface (using PHPStorm you can generate it => https://www.jetbrains.com/help/phpstorm/extract-interface.html)

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (mickaelandrieu) [inline src/Core/Localization/Pack/Import/LocalizationPackImporter.php:168]: Thinking about it, we could use a trait (ok, I don't like traits but...).

Don't change it for now, I need time to decide if we introduce a new trait :)

File: src/Adapter/Language/LanguageManager.php
Comment (sarjon) [reply src/Adapter/Language/LanguageManager.php]: agree, naming things is hard. :/

File: src/Adapter/Localization/LocalUnitsConfiguration.php
Comment (sarjon) [reply src/Adapter/Localization/LocalUnitsConfiguration.php:71]: sure, i could add it, but what this generic error should say?

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php:46]: Use annotations instead :)

https://github.com/PrestaShop/prestafony-project/blob/master/docs/migration-guide.md#manage-the-security

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php]: use annotations instead :)

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php]: would you mind to rename this function `processForm` please?

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php]: use annotations please

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php:146]: you can use `flashErrors(array $errors)` from `FrameworkBundleAdminController`

File: src/PrestaShopBundle/Entity/Repository/TimezoneRepository.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Entity/Repository/TimezoneRepository.php]: you must implements `PrestaShop\PrestaShop\Core\Repository/RepositoryInterface` (it's already the case ;) )

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: it's invalid to use capital letters here, how about `data-minimum-results-for-search` ?

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: Extract this private function to a final class (and interface) and inject it instead: you will simplify A LOT your form type).

Something like `LocalisationPackChoices` with `all()` public function

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]:
```diff
- data-minimumResultsForSearch
+ data-minimum-results-for-search
```

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: Same update everywhere please :+1:

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: You don't need to inject the world in this formtype :)

What I suggest here
* move every private function to his own "Provider/Repository": each one of them MUST return an array
* so the constructor of this formType become `__construct(array $languageChoices, array $countryChoices, array $currencyChoices, array $currencyChoices, array $timezoneChoices) { /* ... */}`

In fact, it's ok to have private function in form types "if" you don't inject new dependencies only to be used in private functions: your private functions are about internal changes of your class: if you change something not related in it, this probably means this function should not be here :)

File: src/PrestaShopBundle/Resources/config/services/core/localization.yml
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/services/core/localization.yml:1]: :+1:

File: src/PrestaShopBundle/Resources/config/services/core_services.yml
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/services/core_services.yml]: I don't know if you were of it, but you could do

- { resource: core/*.yml }

instead.

Don't change it here, we'll do another pull request to improve all imports later.

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]
Good job, some refactoring needed :)

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php
Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: i dont like it either, but i'm not sure if i can use `data-minimum-results-for-search` because it is not  implemented by this PR.

File: src/Core/Localization/Pack/Import/LocalizationPackImportConfig.php
Comment (sarjon) [reply src/Core/Localization/Pack/Import/LocalizationPackImportConfig.php]: it's a value object, do we need interface for it anyway?

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php:46]: it seems that annotations are not merged into develop yet.

File: classes/LocalizationPack.php
Comment (PierreRambaud) [inline classes/LocalizationPack.php:53]: There is a `libxml_clear_errors`, instead using `@` you can use `libxml_use_internal_errors(true)` : http://php.net/manual/fr/function.libxml-use-internal-errors.php

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (PierreRambaud) [inline src/Core/Localization/Pack/Import/LocalizationPackImporter.php]: Duplicate `$errors`

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (PierreRambaud) [reply src/Core/Localization/Pack/Import/LocalizationPackImporter.php:168]: Or an abstract class it is for `TranslatorAwareType` (I don't like traits too ><)

File: src/Core/Localization/Pack/Loader/AbstractLocalizationPackLoader.php
Comment (PierreRambaud) [inline src/Core/Localization/Pack/Loader/AbstractLocalizationPackLoader.php]: If you do nothing in if maybe use ternary  
```php
return false === $xml ? null : $xml;
```

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (PierreRambaud) [inline src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: Hum, what is the result if we are in version `1.12.1.0`?

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php:46]: yes, we will maybe merge 1.7.4.x into develop tomorrow :) (I need to)

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php
Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: ok, we'll change it in another PR /c @eternoendless

Comment (PierreRambaud) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: Both of you are right, you can implement camel case after `data-` and you can use `data-minimum-results-for-search` but agree with @mickaelandrieu it's weird but (again) jquery.select2 does not implement the opts `minimum-results-for-search`, lile ProductFeature.php, ProductPrice.php you must use `data-minimumResultsForSearch` to make it work in the BO.

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Import/LocalizationPackImporter.php]: wow! how i could miss it o_O

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php
Comment (PierreRambaud) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: Variables are useless `$choices[(string) $pack->iso] = (string) $pack->name;` and not sure casting to string is useful

File: classes/LocalizationPack.php
Comment (sarjon) [reply classes/LocalizationPack.php:53]: yeah, but maybe this improvement should be part of another PR?

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (sarjon) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: well, it definitely wouldnt work.
however, localization API now supports 2 digits only (16 for 1.6.x, 17 for 1.7.x) and 1.10+ Prestashop is not coming any time soon (or maybe even never :smiley: )  so i think we should keep same logic as it was in legacy code. what do you think?

File: src/Adapter/Localization/LocalUnitsConfiguration.php
Comment (mickaelandrieu) [reply src/Adapter/Localization/LocalUnitsConfiguration.php:71]: I don't know, I'll ask @Louise asap: don't change anything, we'll improve this behavior later :)

File: src/Core/Localization/Pack/Import/LocalizationPackImportConfig.php
Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Import/LocalizationPackImportConfig.php]: yes because making this class final forbid the modification, so we must alllow developers to provide their own.

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Import/LocalizationPackImporter.php:168]: this is probably the worst use case to use mother/child class in OOP: extends a class when you have nothing related between child and his parent.

So, no no no ^^

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php
Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/ImportLocalizationPackType.php]: sad story... would you mind to create an issue into prestafony project so I can plan a refactoring at some point.

File: classes/LocalizationPack.php
Comment (mickaelandrieu) [reply classes/LocalizationPack.php:53]: Agree this will change the behavior and I don't want the migration blocked by an issue not "strictly" related to it.

The good thing is now we have class with only one purpose, it will be easy to improve each part, including your interesting suggestion @PierreRambaud

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (PierreRambaud) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: @eternoendless @mickaelandrieu Wdty? Could be a surprise bug in the future :/

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php:146]: the problem with `flashErrors(array $errors)` is that it adds errors like:
`$this->addFlash('error', $this->trans($error['key'], $error['domain'], $error['parameters']));`
 in my case, i already have translated errors.

What do you think if we updated `flashErrors(array $errors)` to this:

```php
foreach ($errorMessages as $error) {
    if (is_string($error)) {
        $message = $error;
    else {
        $message = $this->trans($error['key'], $error['domain'], $error['parameters']);
    }
    $this->addFlash('error', $message);
}
```

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php
Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: hmm, should we have some generic interface under `PrestaShopBundle\Form\Admin` namespace like `ChoiceValueProviderInterface` with `getChoices(): array` method so it could be reused multiple times (in this PR i need 5 different choice providers).  what do you think @mickaelandrieu ?

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (PierreRambaud) [reply src/Core/Localization/Pack/Import/LocalizationPackImporter.php:168]: They are sharing translation love :heart:  (:trollface:)

**Comment by mickaelandrieu** [comment]
> it's the same method as above, maybe refactoring by create a method setActive($status);, is it normal to settings this value to 1 or 0 and not to true or false?

Sadly, I think it's a "tiny int" stored in database: I don't like it neither :/

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php
Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: no, because I don't see why values of the form should be part of the contract.... and in this case, which "choices" should be in "getChoices"?  $languageChoices, $countryChoices, $currencyChoices or $timezoneChoices?

We usually rely on interface to protect an implementation from his dependencies, but here we inject the dependencies so we are already protected: we can already type hint on ``array``.  Am I missing something here?

Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: Ok, maybe you are right, but here's and example in code of what I meant. :)

Example choices provider for language. 
```php
final class LanguageChoicesProvider implements ChoiceValueProviderInterface {
    public function __construct(/** some dependencies needed to get language choices */) {
        //...
    }

    public function getChoices(): array {
        // logic to build choices array
    }
}
```
How it would look in yaml form_type definition. We only pass arrays to form type instance.
```yaml
    form.type.localization_configuration:
        class: 'PrestaShopBundle\Form\Admin\Improve\International\Localization\LocalizationConfigurationType'
        arguments:
            - '@prestashop.adapter.legacy.context'
            - '@=service("prestashop.bundle.form.language_choice_provider").getChoices()'
            - '@=service("prestashop.bundle.form.country_choice_provider").getChoices()'
            - '@=service("prestashop.bundle.form.currency_choice_provider").getChoices()'
            - '@=service("prestashop.bundle.form.timezone_choice_provider").getChoices()'
        tags:
            - { name: form.type }
```

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: Oh! interesting :) This can formalize a constraint on ``getChoices`` ... can you add an issue on prestafony-project with this proposal? I have the feeling we should discuss it, but not in this pull request.

Still, I'm not sure *yet* of the gain to introduce the interface: does it make the code more testable, easier to understand or to extend? Are these arrays need to be strictly called "choices" or can they be used in other contexts?

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: I have the feeling that someone has already done that... isn't @michaelKaefer? :)

=> https://github.com/PrestaShop/PrestaShop/pull/9154/files#diff-4f0928deabb59e9c6fd5e5c675628d61R83

Does it fix the bug you are discussing?

**Comment by PierreRambaud** [comment]
> > it's the same method as above, maybe refactoring by create a method setActive($status);, is it normal to settings this value to 1 or 0 and not to true or false?
> 
 >Sadly, I think it's a "tiny int" stored in database: I don't like it neither :/

In fact, MySQL boolean field is a tinyint(1) (YEAH FUCK LOGIC), so we can use true and false that should be automatically converted to tinyint :)
```sql
CREATE TABLE `test_prestashop`.`test_boolean` ( `test` BOOLEAN NOT NULL ) ENGINE = InnoDB;

INSERT INTO `test_boolean` (`test`) VALUES ('1');
INSERT INTO `test_boolean` (`test`) VALUES (1);
INSERT INTO `test_boolean` (`test`) VALUES (TRUE);
```

The dump:
```sql
CREATE TABLE `test_boolean` (
  `test` tinyint(1) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

This is one reason why I hate MySQL :trollface:

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (sarjon) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: not really, in this case `http://api.prestashop.com/localization/{psVersion}/{countryIso}.xml` is used to get localization pack data. `{psVersion}` is passed as `17` and not as `1.7` (same for 1.6, it goes as `16` value).
In legacy part it used to take first 2 string letters from version, so i kept it that way, should we change it?

p.s. maybe this major version  method could be extracted into separate service as it could be reused elsewhere?

File: src/Core/Localization/Pack/Import/LocalizationPackImporter.php
Comment (mickaelandrieu) [inline src/Core/Localization/Pack/Import/LocalizationPackImporter.php]: not need to declare this variable if you dont use it :)

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php]: same here, you don't need the variables... I guess your intent is to describe the function :)

Once these function will be moved to the providers, you won't need it anymore :)

**Comment by sarjon** [comment]
@mickaelandrieu i've addressed most of the comments from the review except some that are still in discussion. :)

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (michaelKaefer) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: @mickaelandrieu @sarjon So the existing method could be used (just replace the "." with an empty string) but it is definitly in the wrong place (a controller). What would be the correct place? Should I create a service for retrieving the actual PS version (in various formats like "1.7.4.1" or "1.7" or "17") like sarjon suggested? I am definitly motivated to do it if you tell me the correct place where to put the code :+1:

File: src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php
Comment (mickaelandrieu) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/LocalizationController.php:146]: why your errors messages are already translated? 

hmm, this is not supposed to happens ^^ I'll think about it, answer tomorrow ;)

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: sure! In `Core/Foundation/Version` class (you may need to create it).

I'll complete an issue in prestafony project with some hints to help you, there is no hurry here ;)

Comment (mickaelandrieu) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: @sarjon, this task is not required in this contribution I consider your work done after a final rebase (I've just merged a migration, so the Link class needs to be updated).

**Review by mickaelandrieu — DISMISSED** [review]

File: src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php
Comment (michaelKaefer) [reply src/Core/Localization/Pack/Loader/RemoteLocalizationPackLoader.php]: @mickaelandrieu Ok, thanks! I'll write you :+1:

**Review by mickaelandrieu — APPROVED** [review]

File: src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationPackChoiceProvider.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationPackChoiceProvider.php:37]: for the record, we are discussing the introduction of a new interface for that: this is why this class is not final at the moment

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by sarjon** [comment]
rebased.

**Comment by sarjon** [comment]
@mickaelandrieu does this need QA approval again? its rebased and uses `Version` service.

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Thanks @sarjon and thanks everyone for your reviews :+1:

**Pull request merged**
